README.org

#+TITLE: Parinfer Mode
#+AUTHOR: DogLooksGood

[[https://melpa.org/#/parinfer][file:https://melpa.org/packages/parinfer-badge.svg]] [[https://stable.melpa.org/#/parinfer][file:https://stable.melpa.org/packages/parinfer-badge.svg]]

[[file:images/logo.png]]

* Demo
[[file:images/demo.gif]]

* Table of contents
- [[#what-is-parinfer][What Is Parinfer]]
- [[#installation][Installation]]
- [[#indent-and-paren-modes][Indent And Paren Modes]]
- [[#credits][Credits]]
- [[#license][License]]

* What Is Parinfer
*Parinfer* is a proof-of-concept editor mode for Lisp programming languages. 
It will infer some changes to keep Parens and Indentation inline with one another.
Parinfer consists of two modes:

- Indent Mode :: Indent Mode gives you full control of indentation, while Parinfer corrects or inserts close-parens
where appropriate.
- Paren Mode :: Paren Mode gives you full control of parens, while Parinfer corrects indentation.

To learn more about Parinfer, please take a look at its [[https://shaunlebron.github.io/parinfer/][home page]].

And the *parinfer-mode* is the implementation on Emacs with [[https://github.com/oakmac/parinfer-elisp][parinfer-elisp]].

* Installation
Parinfer is available on MELPA(and MELPA Stable) now, I recommend to install with [[https://github.com/jwiegley/use-package][use-package]].

Add following code to your ~.emacs~ or ~init.el~.

#+BEGIN_SRC elisp
  (use-package parinfer
    :ensure t
    :bind
    (("C-," . parinfer-toggle-mode))
    :init
    (progn
      (setq parinfer-extensions
            '(defaults       ; should be included.
              pretty-parens  ; different paren styles for different modes.
              evil           ; If you use Evil.
              lispy          ; If you use Lispy. With this extension, you should install Lispy and do not enable lispy-mode directly.
              smart-tab      ; C-b & C-f jump positions and smart shift with tab & S-tab.
              smart-yank))   ; Yank behavior depend on mode.
      (add-hook 'clojure-mode-hook #'parinfer-mode)
      (add-hook 'emacs-lisp-mode-hook #'parinfer-mode)))
#+END_SRC

* Indent And Paren Modes.
Toggle *Indent Mode* and *Paren Mode* with command ~parinfer-toggle-mode~.

* Avoid unexpect buffer changes.
Whenever you switch to *Indent Mode*, parinfer-mode will correct all parens of current buffer.

After opening a file, the parinfer-mode enable *Indent Mode* if it won't make any changes. 
Otherwise, *Paren Mode* will be enabled, you can switch to *Indent Mode* manually later.

Alternatively, you can use ~parinfer-diff~ command to see how will *Indent Mode* modify the buffer with ~Ediff~.

[[file:images/diff_demo.gif]]

And there are some keybindings in Ediff:
| Key  | Description                                               |
|------+-----------------------------------------------------------|
| ~q~  | Quit diff.                                                |
| ~b~  | B->A in Ediff, this can apply change to your origin code. |
| ~ra~ | Restore A in Ediff, this can revert change.               |
| ~n~  | Move to next difference.                                  |
| ~p~  | Move to previous difference.                              |

Also, There's a command ~parinfer-auto-fix~ to fix wrong indentation for whole buffer. (Only works with balanced parens)


* Credits
- [[https://github.com/shaunlebron][shaunlebron]] :: Create [[https://shaunlebron.github.io/parinfer/][Parinfer]].
- [[https://github.com/oakmac][oakmac]] :: Bring Parinfer to Emacs with [[https://github.com/oakmac/parinfer-elisp][parinfer-elisp]].
- [[https://github.com/tumashu][tumashu]] :: Help me a lot in writing this plugin.
- [[https://github.com/purcell][purcell]] & [[https://github.com/syohex][syohex]] :: Advice and Tips for writing emacs plugin

* License
parinferlib.el from [[https://github.com/oakmac/parinfer-elisp][parinfer-elisp]], is licensed under the [[https://github.com/oakmac/parinfer-elisp/blob/master/LICENSE.md][ISC]].

Rest part licensed under the GPLv3.


CHANGELOG.org
#+TITLE: CHANGELOG
#+AUTHOR: DogLooksGood

* v0.3.2
- bugfix.
- ~smart-tab~ extension.

* v0.3.1
- bugfix.
- ~smart-yank~ extension.

* v0.3.0
- bugfix.
- Refactor code.
- Extension module.

* v0.2.0
- Add ~parinfer-skip-command-list~, ~parinfer-skip-command-prefix-list~, ~parinfer-instantly-invoke-command-list~, ~parinfer-instantly-invoke-command-prefix-list~.
- Compatibility with Evil. 
- BUGFIX: Broken ~Package-Requires~. 

* v0.1.4
- Delay processing on large S-expression. customize with ~parinfer-delay-invoke-threshold~ and ~parinfer-delay-invoke-idle~.

* v0.1.3
- Rename all private function with the prefix ~parinfer--~ instead of ~parinfer-~.
- Remove paredit dependencies sine we don't neet it.
- Add group for parinfer-theme.el, fix the compile warnnings ~defface `parinfer-dim-paren-face' fails to specify containing group~.

* v0.1.2
- Fix lag when move cursor in a very large sexp(especially when holding ~C-n~ or ~C-p~).



parinfer-lispy.el
;;; parinfer-lispy.el --- Integration for Parinfer and Lispy

;;; Code

;; -----------------------------------------------------------------------------
;; Fixed commands
;; -----------------------------------------------------------------------------

(defun parinfer-lispy-space ()
  (interactive)
  (if (or (eq (point) (line-beginning-position))
          (eq (point) (line-end-position)))
      (call-interactively 'self-insert-command)
    (progn
      (call-interactively 'self-insert-command)
      (when (parinfer--lispy-left-between-parens-p)
        (backward-char)))))

(defun parinfer-lispy-forward ()
  (interactive)
  (when parinfer--delay-timer
    (parinfer--clean-up))
  (call-interactively 'lispy-forward))

(defun parinfer-lispy-backward ()
  (interactive)
  (when parinfer--delay-timer
    (parinfer--clean-up))
  (call-interactively 'lispy-backward))

(defun parinfer--lispy-paren-char-p (c)
  (or (eq c 40)
      (eq c 91)
      (eq c 123)))

(defun parinfer--lispy-left-between-parens-p ()
  (let ((ca (char-after))
        (cb (char-before (- (point) 1))))
    (and (not (parinfer--in-comment-or-string-p))
         (parinfer--lispy-paren-char-p ca)
         (parinfer--lispy-paren-char-p cb))))

(defun parinfer-lispy-parens ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-parens)
    (call-interactively 'self-insert-command)))

(defun parinfer-lispy-brackets ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-brackets)
    (call-interactively 'self-insert-command)))

(defun parinfer-lispy-braces ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-braces)
    (call-interactively 'self-insert-command)))

(defun parinfer--lispy-switch-mode-behaviour (mode)
  (if (eq mode 'indent)
      (lispy-mode 1)
    (lispy-mode -1)))

(defun parinfer--lispy-init ()
  (define-key lispy-mode-map (kbd "(") 'parinfer-lispy-parens)
  (define-key lispy-mode-map (kbd ")") 'self-insert-command)
  (define-key lispy-mode-map (kbd "[") 'parinfer-lispy-brackets)
  (define-key lispy-mode-map (kbd "]") 'self-insert-command)
  (define-key lispy-mode-map (kbd "{") 'parinfer-lispy-braces)
  (define-key lispy-mode-map (kbd "}") 'self-insert-command)
  (define-key lispy-mode-map (kbd ";") 'parinfer-semicolon)
  (define-key lispy-mode-map [remap lispy-kill] 'kill-line)
  (define-key lispy-mode-map [remap lispy-tick] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-tilde] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-newline-and-indent-plain] 'newline)
  (define-key lispy-mode-map [remap lispy-quotes] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-yank] 'parinfer-yank)
  (define-key lispy-mode-map [remap lispy-colon] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-hash] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-hat] 'self-insert-command)
  (define-key lispy-mode-map [remap lispy-delete-backward] 'parinfer-backward-delete-char)
  (define-key lispy-mode-map [remap lispy-space] 'parinfer-lispy-space))

(defun parinfer--lispy-disable-if-enabled ()
  (when (bound-and-true-p lispy-mode)
    (lispy-mode -1)))

;; -----------------------------------------------------------------------------
;; Public APIs
;; -----------------------------------------------------------------------------

(defun parinfer-lispy-rocks ()
  (add-hook 'parinfer-mode-enable-hook #'parinfer--lispy-disable-if-enabled)
  (add-hook 'parinfer-switch-mode-hook #'parinfer--lispy-switch-mode-behaviour)
  (add-hook 'lispy-mode-hook #'parinfer--lispy-init)
  (parinfer-strategy-add 'default '(parinfer-lispy-parens
                                    parinfer-lispy-braces
                                    parinfer-lispy-brackets
                                    parinfer-lispy-space)))

(provide 'parinfer-lispy)
;;; parinfer-lispy.el ends here


parinfer-theme.el
;;; parinfer-theme.el --- prettify parens for parinfer-mode
;; gray close parens for parinfer-mode indent style.

;;; Code

(defgroup parinfer-group
  nil
  "Parinfer customize group."
  :group 'faces)

(defface parinfer-dim-paren-face
   '((((class color) (background dark))
      (:foreground "grey40"))
     (((class color) (background light))
      (:foreground "grey60")))
   "Parinfer dim paren face."
   :group 'parinfer-group)

(defun parinfer--set-dim-parens (mode)
  "Set dim close parens, which is depended MODE."
  (cl-case mode
    (paren (font-lock-remove-keywords
            nil '((")\\|}\\|]" . 'parinfer-dim-paren-face))))
    (indent (font-lock-add-keywords
             nil '((")\\|}\\|]" . 'parinfer-dim-paren-face)))))
  (font-lock-flush))

(provide 'parinfer-theme)
;;; parinfer-theme.el ends here


parinfer-ext.el


;;; parinfer-ext.el --- Extensions of parinfer-mode

;; Copyright (c) 2016, Shi Tianshu

;; Author: Shi Tianshu
;; Homepage: https://github.com/DogLooksGood/parinfer-mode
;; Keywords: Parinfer

;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 3
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Extensions of parinfer

;;; Code:

(require 'parinfer)

(defgroup parinfer-ext
  nil
  "Parinfer customize group."
  :group 'faces)

;; -----------------------------------------------------------------------------
;; Pretty Parens
;; -----------------------------------------------------------------------------

(defun parinfer-pretty-parens:refresh ()
  (if (fboundp 'font-lock-flush)
      (font-lock-flush)
    (when font-lock-mode
      (with-no-warnings
        (font-lock-fontify-buffer)))))

(defface parinfer-pretty-parens:dim-paren-face
   '((((class color) (background dark))
      (:foreground "grey40"))
     (((class color) (background light))
      (:foreground "grey60")))
   "Parinfer dim paren face."
   :group 'parinfer-ext)

(parinfer-define-extension pretty-parens
  "Pretty parens.

Use rainbow-delimiters for Paren Mode, and dim-style parens for Indent Mode."
  :mount
  (require 'font-lock)
  
  :paren
  (font-lock-remove-keywords
   nil '(("[^\\\\]\\()\\|}\\|\\]\\)" 1 'parinfer-pretty-parens:dim-paren-face)))
  (when (fboundp 'rainbow-delimiters-mode)
    (rainbow-delimiters-mode-enable))
  (parinfer-pretty-parens:refresh)

  :indent
  (when (bound-and-true-p rainbow-delimiters-mode)
    (rainbow-delimiters-mode-disable))
  (font-lock-add-keywords
   nil '(("[^\\\\]\\()\\|}\\|\\]\\)" 1 'parinfer-pretty-parens:dim-paren-face)))
  (parinfer-pretty-parens:refresh))

;; -----------------------------------------------------------------------------
;; defaults
;; -----------------------------------------------------------------------------

(defun parinfer-defaults:company-cancel (&ignored)
  "Invoke when company cancelled, ignore IGNORED."
  (parinfer-indent))

(defun parinfer-defaults:company-finish (&ignored)
  "Invoke when company finished, ignore IGNORED. "
  (parinfer--reindent-sexp))

(parinfer-define-extension defaults
  "Basic compatibility fix."
  :indent
  (when (bound-and-true-p company-mode)
    (add-hook 'company-completion-cancelled-hook
              'parinfer-defaults:company-cancel t t)
    (remove-hook 'company-completion-finished-hook
                 'parinfer-defaults:company-finish t))
  :paren
  (when (bound-and-true-p company-mode)
    (add-hook 'company-completion-finished-hook
              'parinfer-defaults:company-finish t t)
    (remove-hook 'company-completion-cancelled-hook
                 'parinfer-defaults:company-cancel t)))

;; -----------------------------------------------------------------------------
;; lispy
;; -----------------------------------------------------------------------------

(defun parinfer-lispy:space ()
  (interactive)
  (if (or (eq (point) (line-beginning-position))
          (eq (point) (line-end-position)))
      (call-interactively 'self-insert-command)
    (progn
      (call-interactively 'self-insert-command)
      (when (parinfer-lispy:paren-left-and-between-parens-p)
        (backward-char)))))

(defun parinfer-lispy:forward ()
  (interactive)
  (when parinfer--delay-timer
    (parinfer--clean-up))
  (call-interactively 'lispy-forward))

(defun parinfer-lispy:backward ()
  (interactive)
  (when parinfer--delay-timer
    (parinfer--clean-up))
  (call-interactively 'lispy-backward))

(defun parinfer-lispy:paren-char-p (c)
  (or (eq c 40)
      (eq c 91)
      (eq c 123)))

(defun parinfer-lispy:newline ()
  (interactive)
  (parinfer-do
    (call-interactively 'newline-and-indent)))

(defun parinfer-lispy:paren-left-and-between-parens-p ()
  (let ((ca (char-after))
        (cb (char-before (- (point) 1))))
    (and (not (parinfer--in-comment-or-string-p))
         (parinfer-lispy:paren-char-p ca)
         (parinfer-lispy:paren-char-p cb))))

(defun parinfer-lispy:parens ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-parens)
    (call-interactively 'self-insert-command)))

(defun parinfer-lispy:brackets ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-brackets)
    (call-interactively 'self-insert-command)))

(defun parinfer-lispy:braces ()
  (interactive)
  (if (region-active-p)
      (call-interactively 'lispy-braces)
    (call-interactively 'self-insert-command)))

(defun parinfer-lispy:init ()
  (when (package-installed-p 'lispy)
    (require 'lispy)
    (define-key parinfer-mode-map (kbd "(") 'parinfer-lispy:parens)
    (define-key parinfer-mode-map (kbd "{") 'parinfer-lispy:braces)
    (define-key parinfer-mode-map (kbd "[") 'parinfer-lispy:brackets)
    (define-key parinfer-mode-map (kbd "d") 'special-lispy-different)
    (define-key parinfer-mode-map (kbd "-") 'special-lispy-ace-subword)
    (define-key parinfer-mode-map (kbd "q") 'special-lispy-ace-paren)
    (define-key parinfer-mode-map (kbd "a") 'special-lispy-ace-symbol)
    (define-key parinfer-mode-map (kbd "c") 'special-lispy-clone)
    (define-key parinfer-mode-map (kbd "A") 'special-lispy-beginning-of-defun)
    (define-key parinfer-mode-map (kbd "w") 'special-lispy-move-up)
    (define-key parinfer-mode-map (kbd "s") 'special-lispy-move-down)
    (define-key parinfer-mode-map (kbd "h") 'special-lispy-left)
    (define-key parinfer-mode-map (kbd "l") 'special-lispy-right)
    (define-key parinfer-mode-map (kbd "k") 'special-lispy-up)
    (define-key parinfer-mode-map (kbd "j") 'special-lispy-down)
    (define-key parinfer-mode-map (kbd "u") 'special-lispy-undo)
    (define-key parinfer-mode-map (kbd "_") 'special-lispy-underscore)
    (define-key parinfer-mode-map (kbd "m") 'special-lispy-mark-list)
    (define-key parinfer-mode-map (kbd "v") 'special-lispy-view)
    (define-key parinfer-mode-map (kbd "M-m") 'lispy-mark-symbol)
    (define-key parinfer-mode-map (kbd "b") 'special-lispy-back)
    (define-key parinfer-mode-map (kbd "f") 'special-lispy-flow)
    (define-key parinfer-mode-map (kbd "f") 'special-lispy-flow)
    (define-key parinfer-mode-map (kbd "e") 'special-lispy-eval)
    (define-key parinfer-mode-map (kbd "o") 'special-lispy-other-mode)
    (define-key parinfer-mode-map (kbd "y") 'special-lispy-occur)
    (define-key parinfer-mode-map (kbd "r") 'special-lispy-raise)
    (define-key parinfer-mode-map (kbd "C-a") 'lispy-move-beginning-of-line)
    (define-key parinfer-mode-map (kbd "g") 'special-lispy-goto)
    (define-key parinfer-mode-map (kbd ">") 'special-lispy-slurp)
    (define-key parinfer-mode-map (kbd "<") 'special-lispy-barf)
    (define-key parinfer-mode-map (kbd "n") 'special-lispy-new-copy)
    (define-key parinfer-mode-map (kbd "SPC" )'parinfer-lispy:space)))
    

(parinfer-define-extension lispy
  "Integration with Lispy."

  :mount
  (require 'eldoc)
  (eldoc-add-command-completions "lispy-" "parinfer-")
  (parinfer-strategy-add 'default
    '(parinfer-lispy:parens
      parinfer-lispy:braces
      parinfer-lispy:brackets
      parinfer-lispy:space))
  (parinfer-strategy-add 'instantly
    '(newline))
  (parinfer-lispy:init))

;; -----------------------------------------------------------------------------
;; Evil
;; -----------------------------------------------------------------------------

(parinfer-define-extension evil
  "Integration with Evil."
  :mount
  (parinfer-strategy-add 'default
    'evil-delete-char)
  (parinfer-strategy-add 'instantly
    '(evil-delete evil-change evil-change-line evil-paste-before evil-paste-after
      evil-delete-line evil-delete-char evil-delete-backward-char evil-substitute
      evil-change-whole-line evil-force-normal-state evil-normal-state
      evil-shift-left evil-shift-right))
  (parinfer-strategy-add 'skip
    '(evil-previous-line evil-forward-char evil-backward-char evil-next-line
      evil-forward-word evil-forward-word-begin evil-backward-word-begin
      evil-backward-end evil-scroll-page-down evil-scroll-up)))

;; -----------------------------------------------------------------------------
;; Smart yank
;; -----------------------------------------------------------------------------

(defun parinfer-smart-yank:paren-yank ()
  (interactive)
  (let ((yank-str nil))
    (with-temp-buffer
      (yank)
      (parinfer-indent-buffer)
      (setq yank-str (buffer-substring-no-properties (point-min) (point-max))))
    (parinfer-paren-run
     (insert yank-str)
     (parinfer--reindent-sexp))))

(defun parinfer-smart-yank:yank ()
  "Yank behaviour depend on current mode(Indent/Paren)."
  (interactive)
  (cl-case (parinfer-current-mode)
    (indent (call-interactively 'parinfer-yank))
    (paren (call-interactively 'parinfer-smart-yank:paren-yank))))

(parinfer-define-extension smart-yank
  "Yank depend on current mode."
  :mount
  (define-key parinfer-mode-map [remap yank] 'parinfer-smart-yank:yank))

;; -----------------------------------------------------------------------------
;; Smart TAB
;; -----------------------------------------------------------------------------

(defconst parinfer-smart-tab:close-paren-regex "\\()\\|]\\|}\\)")

(defvar parinfer-smart-tab:indicator-line nil)
(make-variable-buffer-local 'parinfer-smart-tab:indicator-line)

(defface parinfer-smart-tab:indicator-face
  '((((class color) (background dark))
     (:background "grey40"))
    (((class color) (background light))
     (:background "grey60")))
   "Parinfer Smart TAB indicator."
   :group 'parinfer-ext)

(defun parinfer-smart-tab:clean-not-skip-this-command-p ()
  (and (symbolp this-command)
       (not (eq this-command 'parinfer-smart-tab:dwim-right-or-complete))
       (not (eq this-command 'parinfer-smart-tab:dwim-right))
       (not (eq this-command 'parinfer-smart-tab:dwim-left))
       (not (eq this-command 'parinfer-smart-tab:forward-char))
       (not (eq this-command 'parinfer-smart-tab:backward-char))))

(defun parinfer-smart-tab:clean-indicator-pre ()
  (interactive)
  (when (and parinfer-smart-tab:indicator-line
             (parinfer-smart-tab:clean-not-skip-this-command-p))
    (save-excursion
      (parinfer--goto-line parinfer-smart-tab:indicator-line)
      (remove-text-properties
       (line-beginning-position)
       (line-end-position)
       '(font-lock-face 'parinfer-smart-tab:indicator-face)))))

(defun parinfer-smart-tab:clean-indicator ()
  (interactive)
  (when (and parinfer-smart-tab:indicator-line
             (parinfer-smart-tab:clean-not-skip-this-command-p))
    (if (and (eq (line-number-at-pos) parinfer-smart-tab:indicator-line))
        (save-excursion
          (end-of-line)
          (while (eq (char-before) 32)
            (backward-delete-char 1)))              
      (save-excursion
        (parinfer--goto-line parinfer-smart-tab:indicator-line)
        (when (parinfer--empty-line-p)
          (delete-region (line-beginning-position) (line-end-position)))))
    (setq parinfer-smart-tab:indicator-line nil)))

(defun parinfer-smart-tab:region-x-and-positions ()
  (let ((begin (region-beginning))
        (end (region-end))
        (m major-mode))
    (setq parinfer--region-shifted nil)
    (deactivate-mark)
    (let* ((sexp-begin (save-excursion (goto-char begin)
                                       (parinfer--goto-previous-toplevel)
                                       (point)))
           (text (buffer-substring-no-properties sexp-begin begin))
           (pos-list nil))
      (progn
       (with-temp-buffer
         (insert text)
         (newline-and-indent)
         (parinfer-indent-buffer)
         (funcall m)
         (setq pos-list (parinfer-smart-tab:find-possible-positions))))
      (goto-char begin)
      (back-to-indentation)
      (let ((x (- (point) (line-beginning-position))))
        (goto-char begin)
        (set-mark-command nil)
        (goto-char end)
        (list x pos-list)))))

(defun parinfer-smart-tab:shift (distance)
  "Shift text.  For right, DISTANCE > 0; left, DISTANCE < 0."
  (parinfer-silent
   (when (use-region-p)
     (let ((mark (mark)))
       (save-excursion
         (indent-rigidly (region-beginning)
                         (region-end)
                         distance)
         (push-mark mark t t)
         (setq parinfer--x-after-shift
               (+ parinfer--x-after-shift distance))
         (setq deactivate-mark nil))))))

(defun parinfer-smart-tab:active-line-region ()
  "Auto adjust region so that the shift can work properly."
  (setq parinfer--x-after-shift (- (point) (line-beginning-position)))
  (let* ((begin (region-beginning))
         (end (region-end))
         (new-begin (save-excursion
                      (goto-char begin)
                      (line-beginning-position))))
    (goto-char new-begin)
    (set-mark-command nil)
    (goto-char end)
    (end-of-line)
    (setq deactivate-mark nil)))

(defun parinfer-smart-tab:shift-right ()
  (interactive)
  (if (eq 'indent (parinfer-current-mode))
      (progn
        (parinfer-smart-tab:active-line-region)
        (let* ((x-and-pos-list (parinfer-smart-tab:region-x-and-positions))
               (x (car x-and-pos-list))
               (pos-list (cadr x-and-pos-list))
               (pos-list-1 (-filter (lambda (n) (> n x))
                                    pos-list)))
          (if (not pos-list-1)
              (parinfer-smart-tab:shift (- x))
            (let ((min-x (-min pos-list-1)))
              (parinfer-smart-tab:shift (- min-x x))))
          (setq parinfer--region-shifted t)))
    (progn
      (setq deactivate-mark t)
      (parinfer--reindent-sexp))))

(defun parinfer-smart-tab:shift-left ()
  (interactive)
  (if (eq 'indent (parinfer-current-mode))
      (progn
        (parinfer-smart-tab:active-line-region)
        (let* ((x-and-pos-list (parinfer-smart-tab:region-x-and-positions))
               (x (car x-and-pos-list))
               (pos-list (cadr x-and-pos-list))
               (pos-list-1 (-filter (lambda (n) (< n x))
                                    pos-list)))
          (if (not pos-list-1)
              (parinfer-smart-tab:shift (- (-max pos-list) x))
            (let ((max-x (-max pos-list-1)))
              (parinfer-smart-tab:shift (- max-x x))))
          (setq parinfer--region-shifted t)))
    (progn
      (setq deactivate-mark t)
      (parinfer--reindent-sexp))))

(defun parinfer-smart-tab:at-close-paren-p ()
  (let ((c (char-after)))
    (or (eq c 41)
        (eq c 93)
        (eq c 125))))

(defun parinfer-smart-tab:should-ignore-positions (pos-list)
  (or (not pos-list)
      (= 1 (length pos-list))))

(defun parinfer-smart-tab:find-possible-positions ()
  (save-excursion
    (unless (= 1 (line-number-at-pos))
      (forward-line -1)
      (while (and (or (parinfer--comment-line-p)
                      (parinfer--empty-line-p))
                  (not (= 1 (line-number-at-pos))))
        (forward-line -1))
      (unless (or (parinfer--comment-line-p)
                  (parinfer--empty-line-p))
        (end-of-line)
        (let ((pos-list '(0)))
          (newline-and-indent)
          (add-to-list 'pos-list (- (point) (line-beginning-position)))
          (delete-indentation)
          (backward-char)
          (while (parinfer-smart-tab:at-close-paren-p)
            (newline-and-indent)
            (add-to-list 'pos-list (- (point) (line-beginning-position)))
            (delete-indentation)
            (backward-char))
          (end-of-line)
          (while (eq 32 (char-before)) 
            (backward-delete-char 1))
          (-distinct pos-list))))))

(defun parinfer-smart-tab:mark-positions (pos-list)
  (unless (parinfer-smart-tab:should-ignore-positions pos-list)
    (setq parinfer-smart-tab:indicator-line (line-number-at-pos))
    (let ((ln (car pos-list))
          (current-x (- (point) (line-beginning-position))))
      (delete-region (line-beginning-position) (line-end-position))
      (cl-loop for i from 0 to ln do
               (if (-contains-p pos-list i)
                   (insert (propertize " " 'font-lock-face 'parinfer-smart-tab:indicator-face))
                 (insert " ")))
      (beginning-of-line)
      (if (> current-x ln)
          (progn (end-of-line) (backward-char))
        (forward-char current-x)))))

(defun parinfer-smart-tab:forward-char ()
  (interactive)
  (when (and (not (parinfer--in-comment-or-string-p))
             (parinfer--empty-line-p)
             (not parinfer-smart-tab:indicator-line))
    (when parinfer--delay-timer
      (parinfer--clean-up))
    (let ((pos-list (parinfer-smart-tab:find-possible-positions)))
      (parinfer-smart-tab:mark-positions pos-list)))
  (if (and parinfer-smart-tab:indicator-line
           (eq (line-number-at-pos) parinfer-smart-tab:indicator-line))
      (progn
        (if (eq (point) (line-end-position))
            (beginning-of-line)
          (forward-char))
        (let ((forward-count 0))
          (while (and (not (eq (face-at-point) 'parinfer-smart-tab:indicator-face))
                      (< forward-count 300))
            (if (eq (point) (line-end-position))
                (beginning-of-line)
              (forward-char))
            (setq forward-count (1+ forward-count)))))
    (call-interactively 'forward-char)))

(defun parinfer-smart-tab:backward-char ()
  (interactive)
  (when (and (not (parinfer--in-comment-or-string-p))
             (parinfer--empty-line-p)
             (not parinfer-smart-tab:indicator-line))
    (when parinfer--delay-timer
      (parinfer--clean-up))
    (let ((pos-list (parinfer-smart-tab:find-possible-positions)))
      (parinfer-smart-tab:mark-positions pos-list)))
  (if (and parinfer-smart-tab:indicator-line
           (eq (line-number-at-pos) parinfer-smart-tab:indicator-line))
      (progn
        (if (eq (point) (line-beginning-position))
            (end-of-line)
          (backward-char))
        (let ((backward-count 0))
          (while (and (not (eq (face-at-point) 'parinfer-smart-tab:indicator-face))
                      (< backward-count 300))
            (if (eq (point) (line-beginning-position))
                (end-of-line)
              (backward-char))
            (setq backward-count (1+ backward-count)))))
    (call-interactively 'backward-char)))

(defun parinfer-smart-tab:dwim-right-or-complete ()
  (interactive)
  (if (eq 'paren parinfer--mode)
      (if (bound-and-true-p company-mode)
          (company-indent-or-complete-common)
        (indent-according-to-mode))
    (cond
     ((region-active-p)
      (call-interactively 'parinfer-smart-tab:shift-right))

     ((parinfer--empty-line-p)
      (parinfer-smart-tab:forward-char))

     ((and (bound-and-true-p company-mode)
           (looking-at "\\_>"))
      (company-complete-common))

     (t
      (progn
        (call-interactively 'set-mark-command)
        (activate-mark)
        (call-interactively 'parinfer-smart-tab:shift-right)
        (deactivate-mark))))))

(defun parinfer-smart-tab:dwim-right ()
  (interactive)
  (if (eq 'paren parinfer--mode)
      (indent-according-to-mode)
    (cond
     ((region-active-p)
      (call-interactively 'parinfer-smart-tab:shift-right))

     ((parinfer--empty-line-p)
      (parinfer-smart-tab:forward-char))

     (t
      (progn
        (call-interactively 'set-mark-command)
        (activate-mark)
        (call-interactively 'parinfer-smart-tab:shift-right)
        (deactivate-mark))))))

(defun parinfer-smart-tab:dwim-left ()
  (interactive)
  (when (eq 'indent parinfer--mode)
    (cond
     ((region-active-p)
      (call-interactively 'parinfer-smart-tab:shift-left))

     ((parinfer--empty-line-p)
      (parinfer-smart-tab:backward-char))

     (t
      (progn
        (call-interactively 'set-mark-command)
        (activate-mark)
        (call-interactively 'parinfer-smart-tab:shift-left)
        (deactivate-mark))))))

(parinfer-define-extension smart-tab
  "Smart forward-char & backward-char."
  :mount
  (add-hook 'post-command-hook 'parinfer-smart-tab:clean-indicator t t)
  (add-hook 'pre-command-hook 'parinfer-smart-tab:clean-indicator-pre t t)
  (define-key parinfer-mode-map [remap forward-char] 'parinfer-smart-tab:forward-char)
  (define-key parinfer-mode-map [remap backward-char] 'parinfer-smart-tab:backward-char)
  (define-key parinfer-region-mode-map [remap parinfer-shift-right] 'parinfer-smart-tab:shift-right)
  (define-key parinfer-region-mode-map [remap parinfer-shift-left] 'parinfer-smart-tab:shift-left)
  
  :unmount
  (remove-hook 'post-command-hook 'parinfer-smart-tab:clean-indicator t)
  (remove-hook 'pre-command-hook 'parinfer-smart-tab:clean-indicator-pre t))



(provide 'parinfer-ext)
;;; parinfer-ext.el ends here



          
          


          
